File: sdk/python/aaf/proposals.py
"""``/api/proposals/*`` sub-client (M8.1 — Gated Proposals).

Mirrors :mod:`backend.api.routers.proposals`. State transitions go
through dedicated helpers (``submit`` / ``approve`` / ``reject`` /
``apply`` / ``withdraw``) so callers don't have to remember the
``:action`` URL suffix.

`apply` records ``status="applied"`` + an audit entry. The framework
deliberately does NOT modify files; humans / CI take the diff and
apply it. Same security model the HTTP API uses.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from .models import (
    Proposal,
    ProposalListResponse,
)

if TYPE_CHECKING:  # pragma: no cover
    from .client import AAFClient, AsyncAAFClient


def _qs(params: dict[str, Any]) -> dict[str, Any]:
    return {k: v for k, v in params.items() if v not in (None, "", [])}


def _make_payload(
    *,
    title: str,
    summary: str,
    motivation: str,
    risk_level: str,
    target_paths: list[str] | None,
    diff: str,
    tags: list[str] | None,
    proposer_kind: str,
    proposer_id: str,
    extras: dict[str, Any] | None,
) -> dict[str, Any]:
    return {
        "title": title,
        "summary": summary,
        "motivation": motivation,
        "risk_level": risk_level,
        "target_paths": list(target_paths or []),
        "diff": diff,
        "tags": list(tags or []),
        "proposer_kind": proposer_kind,
        "proposer_id": proposer_id,
        "extras": dict(extras or {}),
    }


class AsyncProposalsAPI:
    """Async sub-client for ``/api/proposals``."""

    def __init__(self, client: AsyncAAFClient) -> None:
        self._client = client

    async def list_all(
        self,
        *,
        status: str | None = None,
        proposer_id: str | None = None,
        tag: str | None = None,
        page: int = 1,
        page_size: int = 50,
    ) -> ProposalListResponse:
        body = await self._client.request_json(
            "GET",
            "/api/proposals",
            params=_qs(
                {
                    "status": status,
                    "proposer_id": proposer_id,
                    "tag": tag,
                    "page": page,
                    "page_size": page_size,
                }
            ),
        )
        return ProposalListResponse.model_validate(body or {})

    async def get(self, proposal_id: str) -> Proposal:
        body = await self._client.request_json("GET", f"/api/proposals/{proposal_id}")
        return Proposal.model_validate(body)

    async def create(
        self,
        *,
        title: str,
        summary: str = "",
        motivation: str = "",
        risk_level: str = "low",
        target_paths: list[str] | None = None,
        diff: str = "",
        tags: list[str] | None = None,
        proposer_kind: str = "human",
        proposer_id: str = "",
        extras: dict[str, Any] | None = None,
    ) -> Proposal:
        payload = _make_payload(
            title=title,
            summary=summary,
            motivation=motivation,
            risk_level=risk_level,
            target_paths=target_paths,
            diff=diff,
            tags=tags,
            proposer_kind=proposer_kind,
            proposer_id=proposer_id,
            extras=extras,
        )
        body = await self._client.request_json(
            "POST", "/api/proposals", json_body=payload
        )
        return Proposal.model_validate(body)

    async def patch(
        self,
        proposal_id: str,
        *,
        update: dict[str, Any],
        notes: str = "",
    ) -> Proposal:
        payload = dict(update)
        if notes:
            payload["notes"] = notes
        body = await self._client.request_json(
            "PATCH", f"/api/proposals/{proposal_id}", json_body=payload
        )
        return Proposal.model_validate(body)

    async def submit(self, proposal_id: str, *, notes: str = "") -> Proposal:
        return await self._transition(proposal_id, "submit", notes)

    async def approve(self, proposal_id: str, *, notes: str = "") -> Proposal:
        return await self._transition(proposal_id, "approve", notes)

    async def reject(self, proposal_id: str, *, notes: str = "") -> Proposal:
        return await self._transition(proposal_id, "reject", notes)

    async def apply(self, proposal_id: str, *, notes: str = "") -> Proposal:
        """Stamp ``status="applied"``. Does NOT modify files (M8.1 contract)."""
        return await self._transition(proposal_id, "apply", notes)

    async def withdraw(self, proposal_id: str, *, notes: str = "") -> Proposal:
        return await self._transition(proposal_id, "withdraw", notes)

    async def delete(self, proposal_id: str) -> None:
        await self._client.request_json("DELETE", f"/api/proposals/{proposal_id}")

    async def _transition(self, proposal_id: str, action: str, notes: str) -> Proposal:
        body = await self._client.request_json(
            "POST",
            f"/api/proposals/{proposal_id}:{action}",
            json_body={"notes": notes} if notes else None,
        )
        return Proposal.model_validate(body)


class ProposalsAPI:
    """Sync sub-client for ``/api/proposals``."""

    def __init__(self, client: AAFClient) -> None:
        self._client = client

    def list_all(
        self,
        *,
        status: str | None = None,
        proposer_id: str | None = None,
        tag: str | None = None,
        page: int = 1,
        page_size: int = 50,
    ) -> ProposalListResponse:
        body = self._client.request_json(
            "GET",
            "/api/proposals",
            params=_qs(
                {
                    "status": status,
                    "proposer_id": proposer_id,
                    "tag": tag,
                    "page": page,
                    "page_size": page_size,
                }
            ),
        )
        return ProposalListResponse.model_validate(body or {})

    def get(self, proposal_id: str) -> Proposal:
        body = self._client.request_json("GET", f"/api/proposals/{proposal_id}")
        return Proposal.model_validate(body)

    def create(
        self,
        *,
        title: str,
        summary: str = "",
        motivation: str = "",
        risk_level: str = "low",
        target_paths: list[str] | None = None,
        diff: str = "",
        tags: list[str] | None = None,
        proposer_kind: str = "human",
        proposer_id: str = "",
        extras: dict[str, Any] | None = None,
    ) -> Proposal:
        payload = _make_payload(
            title=title,
            summary=summary,
            motivation=motivation,
            risk_level=risk_level,
            target_paths=target_paths,
            diff=diff,
            tags=tags,
            proposer_kind=proposer_kind,
            proposer_id=proposer_id,
            extras=extras,
        )
        body = self._client.request_json("POST", "/api/proposals", json_body=payload)
        return Proposal.model_validate(body)

    def patch(
        self,
        proposal_id: str,
        *,
        update: dict[str, Any],
        notes: str = "",
    ) -> Proposal:
        payload = dict(update)
        if notes:
            payload["notes"] = notes
        body = self._client.request_json(
            "PATCH", f"/api/proposals/{proposal_id}", json_body=payload
        )
        return Proposal.model_validate(body)

    def submit(self, proposal_id: str, *, notes: str = "") -> Proposal:
        return self._transition(proposal_id, "submit", notes)

    def approve(self, proposal_id: str, *, notes: str = "") -> Proposal:
        return self._transition(proposal_id, "approve", notes)

    def reject(self, proposal_id: str, *, notes: str = "") -> Proposal:
        return self._transition(proposal_id, "reject", notes)

    def apply(self, proposal_id: str, *, notes: str = "") -> Proposal:
        """Stamp ``status="applied"``. Does NOT modify files (M8.1 contract)."""
        return self._transition(proposal_id, "apply", notes)

    def withdraw(self, proposal_id: str, *, notes: str = "") -> Proposal:
        return self._transition(proposal_id, "withdraw", notes)

    def delete(self, proposal_id: str) -> None:
        self._client.request_json("DELETE", f"/api/proposals/{proposal_id}")

    def _transition(self, proposal_id: str, action: str, notes: str) -> Proposal:
        body = self._client.request_json(
            "POST",
            f"/api/proposals/{proposal_id}:{action}",
            json_body={"notes": notes} if notes else None,
        )
        return Proposal.model_validate(body)


__all__ = ["AsyncProposalsAPI", "ProposalsAPI"]

File: frontend/src/lib/proposals.ts
/**
 * Gated Proposals API client (M8.1 + P8 Phase C2 bundle additions).
 *
 * Mirrors `backend/api/routers/proposals.py`:
 *
 *   GET    /api/proposals
 *   POST   /api/proposals                              (draft)
 *   GET    /api/proposals/{proposal_id}
 *   PATCH  /api/proposals/{proposal_id}
 *   POST   /api/proposals/{proposal_id}:submit
 *   POST   /api/proposals/{proposal_id}:approve
 *   POST   /api/proposals/{proposal_id}:reject
 *   POST   /api/proposals/{proposal_id}:apply          (status only)
 *   POST   /api/proposals/{proposal_id}:apply-to-bundle (P8 — writes file)
 *   POST   /api/proposals/{proposal_id}:withdraw
 *   DELETE /api/proposals/{proposal_id}
 *
 * `apply` does NOT modify files. Use `:apply-to-bundle` to actually
 * write a bundle proposal back to the manuscript file (admin only).
 */
import { api } from "@/lib/api";
import type {
  CreateProposalInput,
  Proposal,
  ProposalListResponse,
  ProposalStatus,
  UpdateProposalInput,
} from "@/types/api";

function qs(params: Record<string, unknown>): string {
  const search = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v === undefined || v === null || v === "") continue;
    search.set(k, String(v));
  }
  const s = search.toString();
  return s ? `?${s}` : "";
}

const enc = encodeURIComponent;

export interface ListProposalsParams {
  status?: ProposalStatus;
  proposer_id?: string;
  tag?: string;
  page?: number;
  page_size?: number;
}

async function transition(
  id: string,
  action: "submit" | "approve" | "reject" | "apply" | "withdraw",
  notes?: string,
): Promise<Proposal> {
  return api<Proposal>(`/api/proposals/${enc(id)}:${action}`, {
    method: "POST",
    json: notes ? { notes } : null,
  });
}

export const proposalsApi = {
  list(params: ListProposalsParams = {}): Promise<ProposalListResponse> {
    return api<ProposalListResponse>(
      `/api/proposals${qs(params as Record<string, unknown>)}`,
    );
  },

  get(proposalId: string): Promise<Proposal> {
    return api<Proposal>(`/api/proposals/${enc(proposalId)}`);
  },

  create(payload: CreateProposalInput): Promise<Proposal> {
    return api<Proposal>("/api/proposals", { method: "POST", json: payload });
  },

  patch(proposalId: string, payload: UpdateProposalInput): Promise<Proposal> {
    return api<Proposal>(`/api/proposals/${enc(proposalId)}`, {
      method: "PATCH",
      json: payload,
    });
  },

  submit(proposalId: string, notes?: string): Promise<Proposal> {
    return transition(proposalId, "submit", notes);
  },

  approve(proposalId: string, notes?: string): Promise<Proposal> {
    return transition(proposalId, "approve", notes);
  },

  reject(proposalId: string, notes?: string): Promise<Proposal> {
    return transition(proposalId, "reject", notes);
  },

  apply(proposalId: string, notes?: string): Promise<Proposal> {
    return transition(proposalId, "apply", notes);
  },

  /**
   * Re-apply a bundle proposal's recorded `bundle_after` content back
   * to the manuscript file (P8 Phase C2). Distinct from `apply()`,
   * which only stamps `status="applied"`. This call:
   *   - resolves manuscript via body.manuscript_id ?? proposal.extras.manuscript_id
   *   - 409 if the on-disk file changed since the proposal was drafted
   *     unless `force` is true
   *   - 403 if the manuscript is *linked* (external dir) and the
   *     proposal's risk_level is not "low"
   *   - on success: rewrites the file + patches `proposal.extras` with
   *     applied_to_bundle_at/by/size, but does NOT change status.
   */
  applyToBundle(
    proposalId: string,
    body: { manuscript_id?: string; force?: boolean; notes?: string } = {},
  ): Promise<Proposal> {
    return api<Proposal>(`/api/proposals/${enc(proposalId)}:apply-to-bundle`, {
      method: "POST",
      json: body,
    });
  },

  withdraw(proposalId: string, notes?: string): Promise<Proposal> {
    return transition(proposalId, "withdraw", notes);
  },

  delete(proposalId: string): Promise<void> {
    return api<void>(`/api/proposals/${enc(proposalId)}`, { method: "DELETE" });
  },

  /**
   * P9.4 — manually synthesise a heuristic proposal from the last N
   * successful task records. Replaces the pre-P9 behaviour where every
   * successful run auto-drafted a proposal.
   */
  synthesize(body: {
    workflow?: string;
    max_cases?: number;
    actor?: string;
  } = {}): Promise<Proposal> {
    return api<Proposal>("/api/proposals:synthesize", {
      method: "POST",
      json: body,
    });
  },
};
